src/electron/go_vpn_tunnel.ts

```diff
@@ -0,0 +1,316 @@
+
+import {RoutingDaemon} from './routing_service';
+import {ChildProcessHelper} from './process';
+import {pathToEmbeddedBinary} from './util';
+import {VpnTunnel} from './vpn_tunnel';
+
+const isLinux = platform() === 'linux';
+const isWindows = platform() === 'win32';
+
+const TUN2SOCKS_TAP_DEVICE_NAME = isLinux ? 'outline-tun0' : 'outline-tap0';
+const TUN2SOCKS_TAP_DEVICE_IP = '10.0.85.2';
+const TUN2SOCKS_VIRTUAL_ROUTER_IP = '10.0.85.1';
+const TUN2SOCKS_VIRTUAL_ROUTER_NETMASK = '255.255.255.0';
+
+// Cloudflare and Quad9 resolvers.
+const DNS_RESOLVERS = ['1.1.1.1', '9.9.9.9'];
+
+// Establishes a full-system VPN with the help of Outline's routing daemon and child process
+// outline-go-tun2socks. The routing service modifies the routing table so that the TAP device
+// receives all device traffic. outline-go-tun2socks process TCP and UDP traffic from the TAP
+// device and relays it to a Shadowsocks proxy server.
+//
+// |TAP| <-> |outline-go-tun2socks| <-> |Shadowsocks proxy|
+//
+// In addition to the basic lifecycle of the helper processes, this class restarts tun2socks
+// on network changes if necessary.
+// Follows the Mediator pattern in that none of the "helpers" know anything
+// about the others.
+export class GoVpnTunnel implements VpnTunnel {
+  private readonly routing: RoutingDaemon;
+  private readonly tun2socks: GoTun2socks;
+
+  // Extracted out to an instance variable because in certain situations, notably a change in UDP
+  // support, we need to stop and restart tun2socks *without notifying the client* and this allows
+  // us swap the listener in and out.
+  private tun2socksExitListener?: () => void | undefined;
+
+  // See #resumeListener.
+  private terminated = false;
+
+  private isUdpEnabled = false;
+
+  private readonly onAllHelpersStopped: Promise<void>;
+
+  private reconnectingListener?: () => void;
+
+  private reconnectedListener?: () => void;
+
+  constructor(private config: ShadowsocksConfig, private isAutoConnect: boolean) {
+    this.routing = new RoutingDaemon(config.host || '', isAutoConnect);
+    this.tun2socks = new GoTun2socks(config);
+
+    // This pair of Promises, each tied to a helper process' exit, is key to the instance's
+    // lifecycle:
+    //  - once any helper fails or exits, stop them all
+    //  - once *all* helpers have stopped, we're done
+    const exits = [
+      this.routing.onceDisconnected,
+      new Promise<void>((fulfill) => {
+        this.tun2socksExitListener = fulfill;
+        this.tun2socks.onExit = this.tun2socksExitListener;
+      })
+    ];
+    Promise.race(exits).then(() => {
+      console.log('a helper has exited, disconnecting');
+      this.disconnect();
+    });
+    this.onAllHelpersStopped = Promise.all(exits).then(() => {
+      console.log('all helpers have exited');
+      this.terminated = true;
+    });
+
+    // Handle network changes and, on Windows, suspend events.
+    this.routing.onNetworkChange = this.networkChanged.bind(this);
```

Comment:
This is duplicated code. Inject the rounting manager in the constructor and move that to the calling construction function.

I'm imagining something like:
```ts
function createVpnTunne(config, isAutoConnect) {
  const routing = new RoutingDaemon(config.host || '', isAutoConnect);
  let tunnel: VpnTunnel;
  if (env.NETWORK_STACK == "go") {
    tunnel = new GoVpnTunnel(config);
  } else {
    tunnel = new SSLibvTunnel(config);
  }
  routing.onNetworkChange = tunnel.networkChanged;
  if (os.platform() === 'win32') {
    powerMonitor.on('suspend', tunnel.suspend);
     powerMonitor.on('resume', tunnel.resume);
  }
  return tunnel;
}
```

You need to expose networkChanged, suspend and resume methods.

In general it's not a good idea to do a lot in the constructor. It's better to move logic to factory methods and inject dependencies.

Reply:
It seems that we should actually register the powerMonitor listeners on connect instead (see later comment). In that case we don't need that code in the construction.

Reply:
Good suggestion. Done in aa2fd60.